File: challenges-in-python/subsets/subsets.py
# Solving the "Subsets" Challenge
# Challenge Description:
# Given an integer array nums, return all possible subsets (the power set). 
# The solution set must not contain duplicate subsets.

# A subset is a set that contains all elements of another set, including the empty set.
# Example 1:
# Input: nums = [1,2,3]
# Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]

# implementing the function using recursive approach
# Big O:
# Time complexity: O(n * 2^n) | Space complexity: O(n * 2^n)
def subsets(nums):
    def backtrack(start, path):
        res.append(path[:]) # Add the subset to the result
        for i in range(start, len(nums)):
            path.append(nums[i]) # Add the number into the subset
            backtrack(i+1, path) # Move to the next number
            path.pop() # Backtrack the subset by removing the number from the subset

    res = []
    backtrack(0, [])
    return res

# implementing the function using iterative approach
# Big O:
# Time complexity: O(n * 2^n) | Space complexity: O(n * 2^n)
def subsets(nums):
    res = [[]]
    for num in nums:
        res += [item + [num] for item in res] # Add the number into the subset
    return res

# Test the function
def test_subsets():
    assert subsets([1,2,3]) == [[],[1],[1,2],[1,2,3],[1,3],[2],[2,3],[3]]
    assert subsets([0]) == [[]]

File: challenges-in-rust/src/subsets/subsets.rs
// Solving the "Subsets" Challenge
// Problem Statement
// Given an integer array nums, return all possible subsets (the power set). 
// The solution set must not contain duplicate subsets.

// Big O:
// Time Complexity: O(N * 2^N) | Space Complexity: O(N * 2^N)
pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut result = vec![vec![]]; // initialize the result with an empty vector
    for &num in &nums { // iterate through the input vector
        let mut new_subsets = result.clone(); // clone the result
        for subset in &mut new_subsets { // &mut is used to make the subset mutable so that we can append to it
            subset.push(num); // append the current number to the subset
        }
        result.append(&mut new_subsets); 
    }
    result // return the result
}

// The subsets function takes a vector of integers as input and returns a vector of vectors of integers.
// The result vector is initialized with an empty vector.
// We then iterate through the input vector and clone the result vector.
// We then iterate through the cloned result vector and append the current number to each subset.
// Finally, we append the new subsets to the result vector and return the result.

// Let's test the subsets function with a few test cases:
#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    fn as_set(vv: Vec<Vec<i32>>) -> HashSet<Vec<i32>> {
        vv.into_iter().collect()
    }

    #[test]
    fn test_subsets() {
        let input = vec![1, 2, 5];
        let result = subsets(input);
        let expected = vec![
            vec![],
            vec![1],
            vec![2],
            vec![1, 2],
            vec![5],
            vec![1, 5],
            vec![2, 5],
            vec![1, 2, 5],
        ];
        assert_eq!(as_set(result), as_set(expected));
    }

}